cdk/stack.py

```diff
@@ -0,0 +1,198 @@
+            lambda_code_bucket=lambda_code_bucket,
+            timeout=300,
+            vpc_id=oc_vpc_id,
+            security_group_id=oc_security_group_id,
+            environment = {
+                "OPENCAST_API_USER": opencast_api_user,
+                "OPENCAST_API_PASSWORD": opencast_api_password,
+                "DEFAULT_PUBLISHER": default_publisher,
+                "OVERRIDE_PUBLISHER": override_publisher,
+                "OVERRIDE_CONTRIBUTOR": override_contributor,
+                "OC_WORKFLOW": oc_workflow,
+                "OC_FLAVOR": oc_flavor,
+                "OC_TRACK_UPLOAD_MAX": oc_track_upload_max,
+                "OPENCAST_BASE_URL": oc_base_url,
+                "ZOOM_VIDEOS_BUCKET": recordings_bucket.bucket.bucket_name,
+                "UPLOAD_QUEUE_NAME": queues.upload_queue.queue_name,
+                "DEBUG": "0",
+                "OC_OP_COUNT_FUNCTION": op_counts.function.function_name
+            }
+        )
+
+        # grant uploader function permissions
+        queues.upload_queue.grant_consume_messages(uploader.function)
+        op_counts.function.grant_invoke(uploader.function)
+        # this is required so that the presigned s3 urls generated by
+        # the uploader function will be valid
+        recordings_bucket.bucket.grant_read(uploader.function)
+
+        log_notify = ZipLogNotificationsFunction(self, 'LogNotificationFunction',
+            name=names.LOG_NOTIFICATION_FUNCTION,
+            lambda_code_bucket=lambda_code_bucket,
+            environment={}
+        )
+
+        api = ZipApi(self, "RestApi",
+            on_demand_function=on_demand.function,
+            webhook_function=webhook.function,
+            ingest_allowed_ips=ingest_allowed_ips
+        )
+
+        download_event = ZipEvent(self, "DownloadEvent",
+            function=downloader.function,
+            event_rate=downloader_event_rate
+        )
+
+        uploader_event = ZipEvent(self, "UploadEvent",
+            function=uploader.function,
+            event_rate=uploader_event_rate
+        )
+
+        codebuild_project = ZipCodebuildProject(self, "CodebuildProject",
+            lambda_code_bucket=lambda_code_bucket,
+            project_git_url=project_git_url,
+            policy_resources=[
+                on_demand.function.function_arn,
+                webhook.function.function_arn,
+                downloader.function.function_arn,
+                uploader.function.function_arn,
+                op_counts.function.function_arn,
+                log_notify.function.function_arn
+            ]
+        )
+
+        on_demand.add_monitoring(monitoring)
+        webhook.add_monitoring(monitoring)
+        downloader.add_monitoring(monitoring)
+        uploader.add_monitoring(monitoring)
+        op_counts.add_monitoring(monitoring)
+        log_notify.add_monitoring(monitoring)
+        api.add_monitoring(monitoring)
+        queues.add_monitoring(monitoring)
```

Comment:
I'm imagining adding widgets to a dashboard would work similarly, although our zip dashboard has gotten pretty baroque; it might have to work the opposite way, i.e. the dashboard construct takes the other resource objects as input.